pgb_utils/pgb_utils/bigquery.py

```diff
@@ -0,0 +1,634 @@
+                            )
+    # == None if user chose to abort; else DataFrame or generator of same
+    if objects is None:
+        return
+
+    if dry_run: print('\nFiltering for objects within the given cone.')
+
+    # filter out objects not in the cone and return the rest
+    objects_in_cone = _do_cone_search(objects, center, radius, format, iterator)
+    return objects_in_cone
+
+
+def _do_cone_search(objects: Union[pd.DataFrame,Generator[pd.DataFrame, None, None]],
+                    center: astropy.coordinates.SkyCoord,
+                    radius: astropy.coordinates.Angle,
+                    format: str = 'pandas',
+                    iterator: bool = False,
+                    ) -> Union[str,
+                               pd.DataFrame,
+                               Generator[Union[str, pd.DataFrame], None, None]]:
+    """Apply the cone search filter and return appropriate objects.
+    """
+
+    if iterator:  # objects is a generator, return a generator
+        return _do_cone_search_iterator(objects, center, radius, format)
+
+    else:  # objects is single df
+        return _do_cone_search_all(objects, center, radius, format)
+
+def _do_cone_search_iterator(objects: pd.DataFrame, center: astropy.coordinates.SkyCoord, radius: astropy.coordinates.Angle, format):
+    """Iterate objects, format and yield those that are in the cone.
+
+    Args:
+        objects: DataFrame containing histories of multiple objectIds.
+    """
+    for df in objects:
+        in_cone = _object_is_in_cone(df, center, radius)
+
+        if in_cone:  # format and yield
+            if format == 'json':
+                df['objectId'] = df.objectId  # else metadata is lost
+                object = df.reset_index().to_json()  # str
+            else:
+                object = df
+            yield object
+
+def _do_cone_search_all(objects: pd.DataFrame, center: astropy.coordinates.SkyCoord, radius: astropy.coordinates.Angle, format):
+    """Filter out objects not in the cone, format, and return.
+
+    Args:
+        objects: DataFrame containing histories of multiple objectIds.
+    """
+    gb = objects.groupby(level='objectId')
+    objects_in_cone = gb.filter(lambda df: _object_is_in_cone(df,center,radius))
+    if format == 'json':
+        objects_in_cone = objects_in_cone.reset_index().to_json()  # str
+    return objects_in_cone
+
+def _object_is_in_cone(object: pd.DataFrame, center: astropy.coordinates.SkyCoord, radius: astropy.coordinates.Angle):
+    """Checks whether the object's most recent observation has a position that
+    is within a cone defined by center and radius.
+
+    Args:
+        object: DataFrame containing the history of a single objectId.
+                Required columns: ['jd','ra','dec']
+
+    Returns:
+        in_cone: True if object is within radius of center, else False
+    """
+    # get the SkyCoords of the most recent observation
+    obs = object.loc[object['jd']==object['jd'].max(),:]
```

Comment:
> It would be nice when we have a model where we have DIAObject and DIASrc tables. The DIAObject will already have the ra, dec pre-calculated. Taking the most recent position isn't necessarily what you want. You probably want the position of the highest S/N point.

I see, thank you. This also makes me realize that I probably should have called the "candidates" table something like "DIASources" instead..? I could change that now if we settle on a better name.

Reply:
Yeah, maybe.  I think it's clear that we should at least create something that looks like the DIAObject table and DIASrc tables that we expect to see.

Reply:
I renamed the table "candidates" -> "DIASource"